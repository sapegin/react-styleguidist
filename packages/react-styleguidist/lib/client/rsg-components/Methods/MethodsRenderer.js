import "core-js/modules/es6.object.assign";
import "core-js/modules/es6.function.name";

import React from 'react';
import PropTypes from 'prop-types';
import Markdown from 'rsg-components/Markdown';
import Argument from 'rsg-components/Argument';
import Arguments from 'rsg-components/Arguments';
import Name from 'rsg-components/Name';
import JsDoc from 'rsg-components/JsDoc';
import Table from 'rsg-components/Table';

function _extends() { _extends = Object.assign || function (target) { for (let i = 1; i < arguments.length; i++) { const source = arguments[i]; for (const key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const getRowKey = function getRowKey(row) {
  return row.name;
};

export var columns = [{
  caption: 'Method name',
  // eslint-disable-next-line react/prop-types
  render: function render(_ref) {
    const name = _ref.name;

        
const _ref$tags = _ref.tags;

        
const tags = _ref$tags === void 0 ? {} : _ref$tags;
    return React.createElement(Name, {
      deprecated: !!tags.deprecated
    }, "".concat(name, "()"));
  }
}, {
  caption: 'Parameters',
  // eslint-disable-next-line react/prop-types
  render: function render(_ref2) {
    const _ref2$params = _ref2.params;

        
const params = _ref2$params === void 0 ? [] : _ref2$params;
    return React.createElement(Arguments, {
      args: params
    });
  }
}, {
  caption: 'Description',
  // eslint-disable-next-line react/prop-types
  render: function render(_ref3) {
    const description = _ref3.description;

        
const returns = _ref3.returns;

        
const _ref3$tags = _ref3.tags;

        
const tags = _ref3$tags === void 0 ? {} : _ref3$tags;
    return React.createElement("div", null, description && React.createElement(Markdown, {
      text: description
    }), returns && React.createElement(Argument, _extends({
      block: true,
      returns: true
    }, returns)), React.createElement(JsDoc, tags));
  }
}];
export default function MethodsRenderer(_ref4) {
  const methods = _ref4.methods;
  return React.createElement(Table, {
    columns,
    rows: methods,
    getRowKey
  });
}
MethodsRenderer.propTypes = {
  methods: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
    returns: PropTypes.object,
    params: PropTypes.array,
    tags: PropTypes.object
  })).isRequired
};