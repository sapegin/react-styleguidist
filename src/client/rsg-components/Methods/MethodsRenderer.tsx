import React from 'react';
import PropTypes from 'prop-types';
import Markdown from 'rsg-components/Markdown';
import Argument from 'rsg-components/Argument';
import Arguments from 'rsg-components/Arguments';
import Name from 'rsg-components/Name';
import JsDoc from 'rsg-components/JsDoc';
import Table from 'rsg-components/Table';

const getRowKey = (row: MethodProp): string => row.name;

export const columns = [
	{
		caption: 'Method name',
		// eslint-disable-next-line react/prop-types
		render: ({ name, tags = {} }: MethodProp) => (
			<Name deprecated={!!tags.deprecated}>{`${name}()`}</Name>
		),
	},
	{
		caption: 'Parameters',
		// eslint-disable-next-line react/prop-types
		render: ({ params = [] }: MethodProp) => <Arguments args={params} />,
	},
	{
		caption: 'Description',
		// eslint-disable-next-line react/prop-types
		render: ({ description, returns, tags = {} }: MethodProp) => (
			<div>
				{description && <Markdown text={description} />}
				{returns && <Argument block returns {...returns} />}
				<JsDoc {...tags} />
			</div>
		),
	},
];

export interface MethodProp {
	name: string;
	description?: string;
	returns?: { name: string; [key: string]: any };
	params?: any[];
	modifiers?: string[];
	tags?: Record<string, any>;
}

const MethodsRenderer: React.FunctionComponent<{ methods: MethodProp[] }> = ({ methods }) => (
	<Table columns={columns} rows={methods} getRowKey={getRowKey} />
);

MethodsRenderer.propTypes = {
	methods: PropTypes.arrayOf(
		PropTypes.shape({
			name: PropTypes.string.isRequired,
			description: PropTypes.string,
			returns: PropTypes.object,
			params: PropTypes.array,
			tags: PropTypes.object,
		}).isRequired
	).isRequired,
};

export default MethodsRenderer;
