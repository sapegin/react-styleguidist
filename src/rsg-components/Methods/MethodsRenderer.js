import React, { PropTypes } from 'react';
import Code from 'rsg-components/Code';
import Markdown from 'rsg-components/Markdown';
import Styled from 'rsg-components/Styled';
import Group from 'react-group';
import cx from 'classnames';

const styles = ({ font, border, light, name, type }) => ({
	table: {
		width: '100%',
		borderCollapse: 'collapse',
	},
	tableHead: {
		borderBottom: [[1, border, 'solid']],
	},
	tableBody: {
	},
	row: {
	},
	cell: {
		paddingRight: 15,
		paddingTop: 6,
		verticalAlign: 'top',
		fontFamily: font,
		fontSize: 13,
	},
	cellHeading: {
		paddingRight: 15,
		paddingBottom: 6,
		textAlign: 'left',
		fontFamily: font,
		fontWeight: 'bold',
		fontSize: 13,
	},
	cellDesc: {
		width: '70%',
		paddingLeft: 15,
	},
	required: {
		fontFamily: font,
		fontSize: 13,
		color: light,
	},
	name: {
		fontSize: 13,
		color: name,
	},
	type: {
		fontSize: 13,
		color: type,
	},
});

export function MethodsRenderer({ classes, methods }) {
	function renderRows(methods) {
		const rows = [];
		methods.map((method) => {
			rows.push(
				<tr key={method.name} className={classes.row}>
					<td className={classes.cell}><Code className={classes.name}>{method.name}()</Code></td>
					<td className={classes.cell}>{renderParameters(method)}</td>
					<td className={cx(classes.cell, classes.cellDesc)}>
						{renderDescription(method)}
						{renderReturns(method)}
					</td>
				</tr>
			);
		});
		return rows;
	}

	function renderParameters(prop) {
		const { params } = prop;
		const rows = [];
		params.map((param) => {
			const { description, name, type } = param;
			rows.push(
				<div key={name} className={classes.methodParam}>
					<Code className={classes.name}>{name}</Code>
					{type && ': '}
					{type && <Code className={classes.type}>{type.name}</Code>}
					{description && ' — '}
					{description && <Markdown text={description} inline />}
				</div>
			);
		});
		return rows;
	}

	function renderReturns(prop) {
		const { returns } = prop;
		return returns ? (
			<span>
				Returns{' '}
				<Code className={classes.type}>{returns.type.name}</Code>
				{returns.description && ' — '}
				{returns.description && <Markdown text={returns.description} inline />}
			</span>
		) : false;
	}

	function renderDescription(prop) {
		const { description } = prop;
		return (
			<Group>
				{description && <Markdown text={description} inline />}
			</Group>
		);
	}

	return (
		<table className={classes.table}>
			<thead className={classes.tableHead}>
				<tr className={classes.row}>
					<th className={classes.cellHeading}>Name</th>
					<th className={classes.cellHeading}>Parameters</th>
					<th className={cx(classes.cellHeading, classes.cellDesc)}>Description</th>
				</tr>
			</thead>
			<tbody className={classes.tableBody}>
				{renderRows(methods)}
			</tbody>
		</table>
	);
}

MethodsRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	methods: PropTypes.array.isRequired,
};

export default Styled(styles)(MethodsRenderer);
