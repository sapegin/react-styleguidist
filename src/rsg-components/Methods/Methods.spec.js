import React from 'react';
import { parse } from 'react-docgen';
import { MethodsRenderer } from './MethodsRenderer';

function render(methods) {
	const parsed = parse(`
		import { Component, PropTypes } from 'react';
		export default class Cmpnt extends Component {
			${methods.join('\n')}
			render() {
			}
		}
	`);
	return shallow(<MethodsRenderer methods={parsed.methods} classes={{}} />);
}

it('should render public method', () => {
	const actual = render(['/**\n * Public\n * @public\n */\nmethod() {}']);

	expect(actual).toMatchSnapshot();
});

it('should render parameters', () => {
	const actual = render(['/**\n * Public\n * @public\n * @param {Number} value - Description\n */\nmethod(value) {}']);

	expect(actual).toMatchSnapshot();
});

it('should render returns', () => {
	const actual = render(['/**\n * @public\n * @returns {Number} - Description\n */\nmethod() {}']);

	expect(actual).toMatchSnapshot();
});

