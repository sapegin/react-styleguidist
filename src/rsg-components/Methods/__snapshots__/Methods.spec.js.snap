// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MethodsRenderer should render a table 1`] = `
<Consumer(Table)
  columns={
    Array [
      Object {
        "caption": "Method name",
        "render": [Function],
      },
      Object {
        "caption": "Parameters",
        "render": [Function],
      },
      Object {
        "caption": "Description",
        "render": [Function],
      },
    ]
  }
  getRowKey={[Function]}
  rows={
    Array [
      Object {
        "description": "Public",
        "modifiers": Array [],
        "name": "method",
        "params": Array [],
      },
    ]
  }
/>
`;

exports[`PropsRenderer should render JsDoc tags 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Consumer(Name)
        deprecated={false}
      >
        Foo()
      </Consumer(Name)>
    </div>
    <div
      key="1"
    >
      <Consumer(Arguments)
        args={Array []}
      />
    </div>
    <div
      key="2"
    >
      <div>
        <JsDoc
          since={
            Array [
              Object {
                "description": "1.0.0",
                "title": "since",
              },
            ]
          }
        />
      </div>
    </div>
  </li>
</ul>
`;

exports[`PropsRenderer should render deprecated JsDoc tags 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Consumer(Name)
        deprecated={true}
      >
        Foo()
      </Consumer(Name)>
    </div>
    <div
      key="1"
    >
      <Consumer(Arguments)
        args={Array []}
      />
    </div>
    <div
      key="2"
    >
      <div>
        <JsDoc
          deprecated={
            Array [
              Object {
                "description": "Use *another* method",
                "title": "description",
              },
            ]
          }
        />
      </div>
    </div>
  </li>
</ul>
`;

exports[`PropsRenderer should render parameters 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Consumer(Name)
        deprecated={false}
      >
        method()
      </Consumer(Name)>
    </div>
    <div
      key="1"
    >
      <Consumer(Arguments)
        args={
          Array [
            Object {
              "description": "Description",
              "name": "value",
              "optional": undefined,
              "type": Object {
                "name": "Number",
              },
            },
          ]
        }
      />
    </div>
    <div
      key="2"
    >
      <div>
        <Markdown
          text="Public"
        />
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`PropsRenderer should render public method 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Consumer(Name)
        deprecated={false}
      >
        method()
      </Consumer(Name)>
    </div>
    <div
      key="1"
    >
      <Consumer(Arguments)
        args={Array []}
      />
    </div>
    <div
      key="2"
    >
      <div>
        <Markdown
          text="Public"
        />
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`PropsRenderer should render returns 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Consumer(Name)
        deprecated={false}
      >
        method()
      </Consumer(Name)>
    </div>
    <div
      key="1"
    >
      <Consumer(Arguments)
        args={Array []}
      />
    </div>
    <div
      key="2"
    >
      <div>
        <Consumer(Argument)
          block={true}
          description="Description"
          returns={true}
          type={
            Object {
              "name": "Number",
            }
          }
        />
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;
